[#30] Implement AMQP 1.0 transport binding

Alfusainey opened the pull request:
An initial patch implementing the AMQP 1.0 protocol binding for client events. This patch implements the `MessageReader` and `MessageWriter` interfaces to support the reading and writing of cloud events to and from an AMQP representation

@slinkydeveloper I did a demo integration test using a vertx-proton client and server to send and receive cloud event messages. once this is merged, i will provide a sample README

Signed-off-by: Alfusainey Jallow <alf.jallow@gmail.com>

slinkydeveloper commented on amqp/pom.xml:
```diff
@@ -0,0 +1,51 @@
+		<dependency>
+			<groupId>org.apache.qpid</groupId>
+			<artifactId>proton-j</artifactId>
+			<version>0.33.7</version>
```
Make this version a property

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageReader.java:
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2020-Present The CloudEvents Authors
```
AFAIR, this should be 2018-Present, can you check the others too?

Alfusainey replied:
ok, will change accordingly.. i was thinking the start date indicates the date file was created.

slinkydeveloper replied:
It was agreed in past that we should use the same for all the files since the beginning of the project

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageReader.java:
```diff
@@ -0,0 +1,127 @@
+     */
+    @Override
+    protected String toCloudEventsValue(final Object value) {
+        return Objects.requireNonNull(value).toString();
```
Can the attribute be there in the applicationProperties map but with a null value?

Alfusainey replied:
i think this depends on how strict the map implementation used for the applicationProperties is. This reader uses the available HashMap implementation which is less strict and thus allows `null` values (and even null keys). since the `ApplicationProperties` is defined to take any Map type, i added this just to be on the safe side

but i can check if there is already a stricter map implementation for app properties present within proton-j

slinkydeveloper replied:
I think it's fine, my only concern is that, if the value is null, then this method shouldn't be called at all. Maybe you need the null check in `forEachHeader`?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpMessageFactory.java:
```diff
@@ -0,0 +1,103 @@
+
+public class AmqpMessageFactory {
+
+    public static MessageReader createReader(final Message message) {
```
All the public static methods here should be well documented, these are the entrypoints of the module. Also, if you can, add something to root javadoc of the class

Alfusainey replied:
sure 👍

slinkydeveloper commented on amqp/pom.xml:
```diff
@@ -0,0 +1,51 @@
+		<version>2.0.0-SNAPSHOT</version>
+	</parent>
+
+	<artifactId>cloudevents-amqp</artifactId>
```
should we name it following the same pattern of other "integration modules" aka `cloudevents-[protocol]-[library]`? Or better, do we expect someday someone comes in proposing a new amqp integration with another library (eg http integrations)? Or is it proton-j a standard de facto client, pretty much like official java kafka client is the standard de facto client for kafka?

Alfusainey replied:
good question! i dont know the answer TBH. 

i would think of qpid-proton more as a library that can be integrated in any AMQP client/broker/etc. One can use a [`JMS`](https://www.oracle.com/technical-resources/articles/java/intro-java-message-service.html) client, or the rabbit-mq client and there might be other clients too. these clients may not rely on proton-j though.

so i think we can use `cloudevents-[protocol]-[library]`, WDYT?

slinkydeveloper replied:
yeah let's go with `cloudevents-amqp-proton` ok?

Alfusainey replied:
sure, should i then change the name of the class to something with `proton` in the name? e.g `ProtonBasedAmqpBinaryMessageWriter`?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageReader.java:
```diff
@@ -0,0 +1,127 @@
+/*
```
Can we have this class under a subpackage `impl`?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageWriter.java:
```diff
@@ -0,0 +1,104 @@
+/*
```
Can we have this class under a subpackage `impl`?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpConstants.java:
```diff
@@ -0,0 +1,51 @@
+/*
```
Can we have this class under a subpackage `impl`?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpMessageFactory.java:
```diff
@@ -0,0 +1,103 @@
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+
+public class AmqpMessageFactory {
```
Make it public final and add private constructor, also annotate with @ParametersAreNonnullByDefault

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpMessageFactory.java:
```diff
@@ -0,0 +1,103 @@
+     *         name.
+     */
+    @SuppressWarnings("unchecked")
+    private static <T> T getApplicationProperty(final ApplicationProperties props, final String name,
```
Can we have these methods somewhere else like in a class `AmqpUtils` under the subpackage `impl`?

Alfusainey replied:
how about putting them in `AmqpConstants.java`?

slinkydeveloper replied:
That's fine too

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageWriter.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public CloudEventExtensionsWriter withExtension(String name, String value) throws CloudEventRWException {
+        // TODO: for now, extensions are mapped to application-properties
```
WDYM? Do we expect to map them somewhere else?

Alfusainey replied:
this was the [comment](https://github.com/cloudevents/sdk-java/issues/30#issuecomment-723982190) i raised in the issue. 

the spec does not mandate the location where extensions can be stored (and left this to the implementors). your suggestion, as a first pass, was to put them in the application-properties section of the message. but it could be that the extension names may align with the name of amqp message properties and a user of the library would like to store the extensions with the message properties because they align better.

this is why i added this `TODO` note here

slinkydeveloper replied:
I think you can even remove the todo and just leave the comment (even better, add the link to the GH comment where you talked about this)

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpBinaryMessageWriter.java:
```diff
@@ -0,0 +1,104 @@
+    @Override
+    public Message setEvent(EventFormat format, byte[] value) throws CloudEventRWException {
+        message.setContentType(format.serializedContentType());
+        return this.end(new BytesCloudEventData(value));
```
Can you duplicate the code of end here to avoid this wrapping?

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/AmqpMessageFactory.java:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    private static byte[] getPayloadAsByteArray(final Section payload) {
+        if (payload == null) {
+            return null;
+        }
+
+        if (payload instanceof Data) {
+            final Data body = (Data) payload;
+            return body.getValue().getArray();
+        } else if (payload instanceof AmqpValue) {
+            final AmqpValue body = (AmqpValue) payload;
+            if (body.getValue() instanceof byte[]) {
+                return (byte[]) body.getValue();
+            } else if (body.getValue() instanceof String) {
+                return ((String) body.getValue()).getBytes(StandardCharsets.UTF_8);
+            }
+        }
+
+        return null;
+    }
```
After we merge this one, it might be interesting to investigate if we can leverage the `CloudEventData` abstraction to perform some optimizations here (eg when you receive an event from amqp, change something in the attributes, and then send it back to amqp)

slinkydeveloper replied:
Pretty much like we did for json https://github.com/cloudevents/sdk-java/pull/251

Alfusainey replied:
yes, this makes a lot of sense

Alfusainey replied:
@slinkydeveloper created #276 to track this

slinkydeveloper reviewed (commented):
This PR looks reaaaally nice and I'm super excited about it, thanks for your contribution!!!

I left a bunch of questions and style comments. Sorry in advance for the comments related to my poor knowledge of amqp :smile:

matzew commented:
REally nice work! Glad to see this coming!

Alfusainey commented:
i updated the PR to incorporate the comments. i also renamed the class names to include `Proton` so that it is clear which amqp library the implementation uses

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/ProtonBasedAmqpMessageFactory.java:
```diff
@@ -0,0 +1,106 @@
+ * {@link <a href="https://github.com/apache/qpid-proton-j">Qpid Proton</a>}.
+ */
+@ParametersAreNonnullByDefault
+public final class ProtonBasedAmqpMessageFactory {
```
I think without Based is fine, let's not be too much verbose :smile: Same for other classes

Alfusainey replied:
sure 😄

slinkydeveloper commented on amqp/src/test/java/io/cloudevents/amqp/ProtonBasedAmqpMessageWriterTest.java:
```diff
@@ -0,0 +1,111 @@
+        assertThat(actualMessage.getApplicationProperties()).isNull();
+    }
+
+    private Message translateBinaryEvent(final CloudEvent event) {
```
Can you try a message with cloudevents extensions too?

Alfusainey replied:
i notice a small bug(?) while testing this. 

In `io.cloudevents.SpecVersion`, the `datacontentencoding` attribute name is included in the list of optional attributes for the V03 (i.e in CloudEventAttributes#getAttributeNames()). However, when i use that name to retrieve its value using CloudEventV03#getAttribute(String), i get an `IllegalArgumentException:  Spec version v0.3 doesn't have attribute named datacontentencoding`;

i think an additional branch should be added to the switch statement to handle the case when the attributeName is `datacontentencoding`. Since the value of this attribute is not stored inside a CloudEventV03 object, we can just do nothing in this case, instead of raising an exception. any thoughts, @slinkydeveloper ? this is what am doing now to make my tests happy 😄

slinkydeveloper replied:
The `datacontentencoding` is a funny one :smile: In the 0.3 spec, it was a bit unclear its role outside the json itself. So in this sdk, I decided to don't save it in the event, because it doesn't make sense to do so. It is in fact only used in the json format, when the event is marshalled/unmarshalled back and forth the wire.

I guess the right behaviour here is that, when `CloudEventV03#getAttribute(String)` is invoked with `datacontentencoding`, it should just return null always.

slinkydeveloper replied:
I can provide a patch for it, one moment

slinkydeveloper replied:
Here it is https://github.com/cloudevents/sdk-java/pull/279 Mind doing a review?

Alfusainey replied:
ok, done

slinkydeveloper reviewed (commented):
Small nits and we can merge

slinkydeveloper commented:
@Alfusainey I'm looking now at the errors and it seems they're related to this pr, let me show you

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/impl/ProtonAmqpBinaryMessageReader.java:
```diff
@@ -0,0 +1,131 @@
+     * Create an instance of an AMQP message reader.
+     *
+     * @param version                  The version of the cloud event message.
+     * @param applicationPropereties   The application properties of the AMQP message that contains
```
Typo in applicationProperties

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/impl/ProtonAmqpBinaryMessageReader.java:
```diff
@@ -0,0 +1,131 @@
+     * @param applicationPropereties   The application properties of the AMQP message that contains
+     *                                 the cloud event metadata (i.e attributes and extensions).
+     *                                 The applicationProperties MUST not be {@code null}.
+     * @param contentType              The content-type property of the AMQP message or {@null} if the message content type is unknown.
```
s/{@null}/{@code null}

Alfusainey replied:
thanks for catching 👍

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/ProtonAmqpMessageFactory.java:
```diff
@@ -0,0 +1,106 @@
+import io.cloudevents.rw.CloudEventWriter;
+/**
+ * A factory class providing convenience methods for creating MessageReader and MessageWriter instances based on
+ * {@link <a href="https://github.com/apache/qpid-proton-j">Qpid Proton</a>}.
```
You don't need to wrap this in a link

Alfusainey commented:
> @Alfusainey I'm looking now at the errors and it seems they're related to this pr, let me show you

@slinkydeveloper the build is actually failing to generate javadoc for the cloudevents-api module. here is a summary:

```
[ERROR] Exit code: 1 - javadoc: error - The code being documented uses modules but the packages defined in http://docs.oracle.com/javase/8/docs/api/ are in the unnamed module.
[ERROR] sdk-java/api/src/main/java/io/cloudevents/CloudEvent.java:31: warning: no @return
[ERROR]     CloudEventData getData();
[ERROR]                    ^
[ERROR] sdk-java/api/src/main/java/io/cloudevents/rw/CloudEventReader.java:38: warning: no @param for <V>
[ERROR]     default <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory) throws CloudEventRWException {
[ERROR]                                                  ^
[ERROR] sdk-java/api/src/main/java/io/cloudevents/rw/CloudEventReader.java:38: warning: no @param for <R>
[ERROR]     default <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory) throws CloudEventRWException {
[ERROR]                                                  ^
[ERROR] sdk-java/api/src/main/java/io/cloudevents/rw/CloudEventReader.java:38: warning: no @return
[ERROR]     default <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory) throws CloudEventRWException {
[ERROR]                                                  ^
[ERROR] sdk-java/api/src/main/java/io/cloudevents/rw/CloudEventReader.java:45: warning: no @param for <V>
[ERROR]     <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory, @Nullable CloudEventDataMapper mapper) throws CloudEventRWException;
[ERROR]                                          ^
[ERROR] sdk-java/api/src/main/java/io/cloudevents/rw/CloudEventReader.java:45: warning: no @param for <R>
[ERROR]     <V extends CloudEventWriter<R>, R> R read(CloudEventWriterFactory<V, R> writerFactory, @Nullable CloudEventDataMapper mapper) throws CloudEventRWException;
```
any idea what am doing wrong?

Alfusainey commented on amqp/src/test/java/io/cloudevents/amqp/ProtonAmqpMessageWriterTest.java:
```diff
@@ -21,43 +21,47 @@
+        try {
+            expectedMessage = translateBinaryEvent(binaryEvent);
+        } catch (final Exception e) {
+            System.out.println(e);
```
i used this for debugging but forgot to remove. will include this in the final update

Alfusainey commented:
i needed to modify the maven-javadoc-plugin configuration for the build to work for me (https://stackoverflow.com/a/61884267/8462878). i will not commit this change though :-)

slinkydeveloper commented:
> @slinkydeveloper the build is actually failing to generate javadoc for the cloudevents-api module

So on your machine javadoc tool fails for warnings? Interesting... Anyway, we plan to fix javadoc properly soon https://github.com/cloudevents/sdk-java/issues/264 so hopefully this will be fixed asap.

slinkydeveloper commented:
@Alfusainey thanks again for this contribution :rocket:

Alfusainey commented:
> @Alfusainey thanks again for this contribution 🚀

you're welcome, and thanks a lot for taking the time to review it! I will be hanging around this project from now on 😄

slinkydeveloper commented:
Oh yeah, feel free to check out the other good first issues https://github.com/cloudevents/sdk-java/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22 and also join the cncf slack if you want to https://cloud-native.slack.com/archives/CCXF3CBL1

slinkydeveloper commented on amqp/src/main/java/io/cloudevents/amqp/ProtonAmqpMessageFactory.java:
```diff
@@ -0,0 +1,105 @@
+     * @return            A message reader capable of representing a CloudEvent from
+     *                    a message <em>content-type</em> property and <em>application-data</em>.
+     */
+    public static MessageReader createReader(final String contentType, final byte[] payload) {
```
Hi @Alfusainey I just noticed this method and i was wondering... do we really need it? Can you provide a use case for it?
